File: src/test/java/ar/edu/unq/spring/controller/helper/MockMVCInventarioController.java
package ar.edu.unq.spring.controller.helper;

import ar.edu.unq.spring.controller.dto.ItemDTO;
import ar.edu.unq.spring.controller.dto.PersonajeDTO;
import ar.edu.unq.spring.modelo.Item;
import ar.edu.unq.spring.modelo.Personaje;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.ServletException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;

import java.util.Collection;
import java.util.List;

// Con @Component le decimos a Spring que inicialice este objeto y lo agregue a su contexto
// Por lo que queda disponible para ser Autowired en otros lugares
@Component
public class MockMVCInventarioController {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    // MockMVC por default re-throwea "NestedServletException" cuando le llega un error.
    // Aca estamos esperando por esa excepcion, y re-throweamos la causa.
    // De esa forma, podemos testear tranquilos y esperar las excepciones de negocio y applicacion del otro lado.
    public ResultActions performRequest(MockHttpServletRequestBuilder requestBuilder) throws Throwable {
        try {
            return mockMvc.perform(requestBuilder);
        } catch (ServletException e) {
            throw e.getCause();
        }
    }

    public Long guardarItem(Item item) throws Exception {
        return guardarItem(item, HttpStatus.CREATED);
    }

    private Long guardarItem(Item item, HttpStatus expectedStatus) throws Exception {
        var dto = ItemDTO.desdeModelo(item);
        var json = objectMapper.writeValueAsString(dto);

        return Long.parseLong(
                mockMvc.perform(MockMvcRequestBuilders.post("/inventario")
                                .contentType(MediaType.APPLICATION_JSON)
                                .content(json))
                        .andExpect(MockMvcResultMatchers.status().is(expectedStatus.value()))
                        .andReturn().getResponse().getContentAsString()
        );
    }

    public Collection<Item> allItems() throws Throwable {
        var json = performRequest(MockMvcRequestBuilders.get("/inventario/all"))
                .andExpect(MockMvcResultMatchers.status().isOk())
                .andReturn().getResponse().getContentAsString();

        Collection<ItemDTO> dtos = objectMapper.readValue(
                json,
                objectMapper.getTypeFactory().constructCollectionType(List.class, ItemDTO.class)
        );

        return dtos.stream().map(ItemDTO::aModelo).toList();
    }

    public void recoger(Long personajeId, Long itemId, HttpStatus expectedStatus) throws Throwable {
        performRequest(MockMvcRequestBuilders.put("/inventario/personaje/" + personajeId + "/recoger/item/" + itemId))
                .andExpect(MockMvcResultMatchers.status().is(expectedStatus.value()));
    }

    public Item heaviestItem() throws Throwable {
        var json = performRequest(MockMvcRequestBuilders.get("/inventario/itemMasPesado"))
                .andExpect(MockMvcResultMatchers.status().isOk())
                .andReturn().getResponse().getContentAsString();

        var dto = objectMapper.readValue(json, ItemDTO.class);
        return dto.aModelo();
    }

    public Collection<Item> getMasPesados(int peso) throws Throwable {
        var json = performRequest(MockMvcRequestBuilders.get("/inventario/masPesados/" + peso))
                .andExpect(MockMvcResultMatchers.status().isOk())
                .andReturn().getResponse().getContentAsString();

        Collection<ItemDTO> dtos = objectMapper.readValue(
                json,
                objectMapper.getTypeFactory().constructCollectionType(List.class, ItemDTO.class)
        );

        return dtos.stream().map(ItemDTO::aModelo).toList();
    }

    public Collection<Item> getItemsPersonajesDebiles(int vida) throws Throwable {
        var json = performRequest(MockMvcRequestBuilders.get("/inventario/itemsPersonajesDebiles/" + vida))
                .andExpect(MockMvcResultMatchers.status().isOk())
                .andReturn().getResponse().getContentAsString();

        Collection<ItemDTO> dtos = objectMapper.readValue(
                json,
                objectMapper.getTypeFactory().constructCollectionType(List.class, ItemDTO.class)
        );

        return dtos.stream().map(ItemDTO::aModelo).toList();
    }
}

File: src/test/kotlin/ar/edu/unq/spring/service/helper/MockMVCInventarioController.kt
package ar.edu.unq.spring.service.helper

import ar.edu.unq.spring.controller.InventarioControllerREST
import ar.edu.unq.spring.controller.dto.ItemJsonDTO
import ar.edu.unq.spring.controller.dto.PersonajeJsonDTO
import ar.edu.unq.spring.modelo.Item
import ar.edu.unq.spring.modelo.Personaje
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.ResultActions
import org.springframework.test.web.servlet.ResultMatcher
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.util.NestedServletException
import java.lang.reflect.Proxy

// Con @Component le decimos a Spring que inicialise este objeto y lo agregue a su contexto
// Por lo que queda disponible para ser Autowired en otro u otros lugares
@Component
class MockMVCInventarioController {
    @Autowired
    lateinit var mockMvc: MockMvc
    @Autowired
    lateinit var objectMapper: ObjectMapper

    // MockMVC por default re-throwea "NestedServletException" cuando le llega un error.
    // Aca estamos esperando por esa excepcion, y re-throweamos la causa.
    // De esa forma, podemos testear tranquilos y esperar las excepciones de negocio y applicacion del otro lado.
    private fun performRequest(requestBuilder: MockHttpServletRequestBuilder): ResultActions {
        return try {
            mockMvc.perform(requestBuilder)
        } catch (e: NestedServletException) {
            throw e.cause ?: e
        }
    }

    fun guardarPersonaje(personaje: Personaje, expectedStatus: HttpStatus = HttpStatus.OK): Long {
        val dto = PersonajeJsonDTO.desdeModelo(personaje)
        val json = objectMapper.writeValueAsString(dto)
        return performRequest(
            post("/inventario/personaje")
                .contentType(MediaType.APPLICATION_JSON).content(json)
        )
            .andExpect(status().`is`(expectedStatus.value()))
            .andReturn().getResponse().getContentAsString().toLong()
    }

    fun guardarItem(item: Item): Long {
        val dto = ItemJsonDTO.desdeModelo(item)
        val json = objectMapper.writeValueAsString(dto)

        return performRequest(
            post("/inventario")
                .contentType(MediaType.APPLICATION_JSON).content(json)
        )
            .andExpect(status().isOk())
            .andReturn().getResponse().getContentAsString().toLong()
    }

    fun recuperarPersonaje(personajeId: Long): Personaje {
        val json = performRequest(get("/inventario/personaje/$personajeId"))
            .andExpect(status().isOk())
            .andReturn().getResponse().getContentAsString()

        val dto = objectMapper.readValue(json, PersonajeJsonDTO::class.java)
        return dto.aModelo()
    }

    fun recuperarItem(itemId: Long): Item {
        val json = performRequest(get("/inventario/$itemId"))
            .andExpect(status().isOk())
            .andReturn().getResponse().getContentAsString()

        val dto = objectMapper.readValue(json, ItemJsonDTO::class.java)
        return dto.aModelo()
    }

    fun recoger(personajeId: Long, itemId: Long, expectedStatus: HttpStatus = HttpStatus.OK) {
        performRequest(put("/inventario/recoger/$personajeId/$itemId"))
            .andExpect(status().`is`(expectedStatus.value()))
    }

    fun allItems(): Collection<Item> {
        val json = performRequest(get("/inventario/allItems"))
            .andExpect(status().isOk())
            .andReturn().getResponse().getContentAsString()

        val dtos: Collection<ItemJsonDTO> = objectMapper.readValue(
            json,
            objectMapper.typeFactory.constructCollectionType(MutableList::class.java, ItemJsonDTO::class.java)
        )
        return dtos.map { dto -> dto.aModelo() }
    }

    fun getMasPesados(peso: Int): Collection<Item> {
        val json = performRequest(get("/inventario/masPesados/$peso"))
            .andExpect(status().isOk())
            .andReturn().getResponse().getContentAsString()

        val dtos: Collection<ItemJsonDTO> = objectMapper.readValue(
            json,
            objectMapper.typeFactory.constructCollectionType(MutableList::class.java, ItemJsonDTO::class.java)
        )
        return dtos.map { dto -> dto.aModelo() }
    }

    fun getItemsPersonajesDebiles(vida: Int): Collection<Item> {
        val json = performRequest(get("/inventario/itemsPersonajesDebiles/$vida"))
            .andExpect(status().isOk())
            .andReturn().getResponse().getContentAsString()

        val dtos: Collection<ItemJsonDTO> = objectMapper.readValue(
            json,
            objectMapper.typeFactory.constructCollectionType(MutableList::class.java, ItemJsonDTO::class.java)
        )
        return dtos.map { dto -> dto.aModelo() }
    }

    fun heaviestItem(): Item {
        val json = performRequest(get("/inventario/heaviestItem"))
            .andExpect(status().isOk())
            .andReturn().getResponse().getContentAsString()

        val dto = objectMapper.readValue(json, ItemJsonDTO::class.java)
        return dto.aModelo()
    }

}